v2/yarpcgrpc/dialer.go

```diff
@@ -21,30 +21,82 @@
 package grpc
 
 import (
-	"go.uber.org/yarpc/api/peer"
+	"context"
+	"sync"
+
+	"go.uber.org/multierr"
+	"go.uber.org/yarpc/v2"
+	"go.uber.org/yarpc/v2/yarpcpeer"
 )
 
-// NewDialer creates a transport that is decorated to retain peers with
-// additional gRPC dial options.
-func (t *Transport) NewDialer(options ...DialOption) *Dialer {
-	return &Dialer{trans: t, options: newDialOptions(options)}
-}
+var _ yarpc.Dialer = (*Dialer)(nil)
 
 // Dialer is a decorator for a gRPC transport that threads dial options for
 // every retained peer.
 type Dialer struct {
-	trans   *Transport
-	options *dialOptions
+	lock          sync.Mutex
+	options       *dialOptions
+	addressToPeer map[string]*grpcPeer
 }
 
-var _ peer.Transport = (*Dialer)(nil)
+// Start starts the gRPC dialer.
+func (d *Dialer) Start(_ context.Context) error {
+	d.addressToPeer = make(map[string]*grpcPeer)
```

Comment:
Should there be a constructor for a `Dialer` so that this is handled elsewhere? If we like the convenience of constructing a `Dialer` in place, we might want to verify that the `addressToPeer` map is nil before we trump its contents.